Kotlin conversion and simplification

pedrodanielcsantos opened the pull request:
This PR adds Kotlin to this project's configuration and converts all models from (boring) Java classes to (much more concise and clear) Kotlin ones.

All classes' and corresponding properties were marked as open to facilitate their extension on consumers of this library.

rastislavvasko commented on src/main/java/com/todoist/pojo/Person.kt:
```diff
@@ -0,0 +1,77 @@
+
+    companion object {
+        /* Matches leading and trailing spaces and special characters commonly found in name fields. */
+        private val ESCAPE_PATTERN = Pattern.compile("^\\s+|\\s+$|[\\(\\)\\.,\\-_\\[\\]'\"]")
```
We could use a raw string to have a cleaner regex.

rastislavvasko commented on src/main/java/com/todoist/pojo/Person.kt:
```diff
@@ -0,0 +1,77 @@
+        @JvmStatic fun getDefaultAvatarText(fullName: String?): String {
+            if (fullName != null) {
+                val names = ESCAPE_PATTERN.matcher(fullName).replaceAll("").split(
+                        "\\s+".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
```
Why the `drop` part? We already split by whitespace, so there shouldn't be any empty parts?

I think we can also make the algorithm below more concise and idiomatic.

pedrodanielcsantos commented on src/main/java/com/todoist/pojo/Collaborator.kt:
```diff
@@ -0,0 +1,79 @@
+) : Person(id, email, fullName, imageId, deleted) {
+
+    /**
+     * @return Unmodifiable set of active project ids.
```
These were legacy comments from the old implementation that slipped through! Removing them 😊

pedrodanielcsantos commented on src/main/java/com/todoist/pojo/Collaborator.kt:
```diff
@@ -0,0 +1,79 @@
+     *
+     * @see {@link .setProjectState
+     */
+    fun setProjectsActive(projectsActive: Collection<Long>) {
```
Because the `setter` method allows for setting any `collection` whereas if we had the property setter as public, we could only set `Set`s, as we can't customize the property setter parameter type. Do you see a better way of doing it ? 🤔

rastislavvasko replied:
But is it useful to allow any collection? I.e. for `Int` properties we don't allow `Number` as arg and then convert them.

pedrodanielcsantos replied:
This was the way it was before, e.g.: https://github.com/Doist/TodoistPojos/blob/master/src/main/java/com/todoist/pojo/Collaborator.java#L76-L78

So I just kept the workflow. I'd say we can keep it for now ? To don't break any existing workflow and then we can re-evaluate ?

rastislavvasko replied:
Okay, let's do that. :)

pedrodanielcsantos commented on src/main/java/com/todoist/pojo/Collaborator.kt:
```diff
@@ -0,0 +1,79 @@
+    }
+
+    /**
+     * Sets the state for a specific project id.
```
Once again, legacy comments from the old implementation 😅 Sorry!

pedrodanielcsantos commented on src/main/java/com/todoist/pojo/Collaborator.kt:
```diff
@@ -0,0 +1,79 @@
+    companion object {
+        const val STATE_ACTIVE = "active"
+        const val STATE_INVITED = "invited"
+        const val STATE_DELETED = "deleted"
```
The previous implementation used strings here. 
I'd say that we could keep strings for now (also as this might affect parsing in the app's side and it's not a kotlin-exclusive feature) and then add `enums` all around in a separate PR (to also try and keep this one more contained, as it's already massive). What do you think ?

rastislavvasko replied:
Yeah, sounds good!

pedrodanielcsantos commented on src/main/java/com/todoist/pojo/Features.kt:
```diff
@@ -0,0 +1,10 @@
+package com.todoist.pojo
+
+open class Features(
+        open var text: String?,
```
This attribute was included in the `Java` version of this model. 
So, given the current PR's scope, I'd maintain it - to keep this PR mostly as a conversion and not doing radical changes - and add an issue for us to revise this later. What do you think ?

rastislavvasko replied:
Ok! 👍

pedrodanielcsantos commented on src/main/java/com/todoist/pojo/User.kt:
```diff
@@ -0,0 +1,30 @@
+package com.todoist.pojo
+
+open class User<TzI: TzInfo, F: Features>(
```
Personally, I like the "pattern" of mimicking the capital letters of a class when using it as a generic type. Still, between option 1 and 2, I prefer 1 so, given that you feel strongly about this, I'll change it and use the first letter only. 👍

pedrodanielcsantos commented on src/main/java/com/todoist/pojo/TzInfo.kt:
```diff
@@ -0,0 +1,9 @@
+package com.todoist.pojo
+
+open class TzInfo(
+        open var text: String?,
```
Replied above. 👍

pedrodanielcsantos commented on src/main/java/com/todoist/pojo/TodoistObject.kt:
```diff
@@ -0,0 +1,15 @@
+package com.todoist.pojo
+
+open class TodoistObject(open var id: Long, open var isDeleted: Boolean) {
+    open fun hasTempId() = id < 0
```
Yeah, I think we should keep it like this, as we use this same approach in other places. 🤔

pedrodanielcsantos commented on src/main/java/com/todoist/pojo/Item.kt:
```diff
@@ -0,0 +1,58 @@
+
+    open var priority: Int = priority
+        /**
+         * Returns the priority within the bounds defined by [.MIN_PRIORITY] and [.MAX_PRIORITY].
```
I would say that it makes sense to maintain it like this because the comment is applicable to the (custom) getter only but the AS detail makes a difference. 🤔

rastislavvasko replied:
Yeah and we can also reword it so that it makes sense as a property-level comment and not just a getter or setter comment.

pedrodanielcsantos commented on src/main/java/com/todoist/pojo/Item.kt:
```diff
@@ -0,0 +1,58 @@
+            }
+        }
+
+    open fun setLabels(labels: Collection<Long>?) {
```
The justification for the public method and private property setter is the same as in here: https://github.com/Doist/TodoistPojos/pull/29#discussion_r292388933.

pedrodanielcsantos commented on src/main/java/com/todoist/pojo/TzInfo.kt:
```diff
@@ -0,0 +1,9 @@
+
+open class TzInfo(
+        open var text: String?,
+        open var timezone: String?,
```
Hum, I don't think so. I left it as nullable "just in case", but it's probably better to have it as non-null.

pedrodanielcsantos commented on src/main/java/com/todoist/pojo/TodoistObject.kt:
```diff
@@ -0,0 +1,15 @@
+        }
+    }
+
+    override fun hashCode() = (id xor id.ushr(32)).toInt()
```
I believe we can keep it as is, it's not super hard to read, so I'd say we can maintain it ... ?

pedrodanielcsantos commented on src/main/java/com/todoist/pojo/Person.kt:
```diff
@@ -0,0 +1,77 @@
+        /* Matches leading and trailing spaces and special characters commonly found in name fields. */
+        private val ESCAPE_PATTERN = Pattern.compile("^\\s+|\\s+$|[\\(\\)\\.,\\-_\\[\\]'\"]")
+
+        private val LIGHT_AVATAR_COLORS = longArrayOf(0XFFE9952C, 0XFFE16B2D, 0XFFD84B40, 0XFFE8435A, 0XFFE5198A,
```
> We could use the experimental UIntclass here instead through uintArrayOf(...).toIntArray()?

If we do that, AS gives an error `Conversion of signed constants to unsigned ones is prohibited`. Do you have any idea to overcome this ? Couldn't think of something obviously better, as Java was much more loose when handling these situations. 🤔

rastislavvasko replied:
Yeah, we need to use unsigned constants (append `u`). 😊

pedrodanielcsantos replied:
🤦‍♂️

pedrodanielcsantos commented:
Thanks a lot for the first thorough review @rastislavvasko! I've either addressed your comments or replied to them. Please let know what you think of this.

Just have one question regarding this: 
> Naming boolean properties: We're mostly using isLoremIpsum rather than loremIpsum, which I think is good, but let's go all in then and do it for everything? E.g. also for Item.checked and Filter.deleted, Features.goldTheme... even though it means being out-of-sync with the API naming.

You mean having _all_ boolean fields as `isLoremIpsum` or all fields as `loremIpsum` ? If I understood correctly, it's the former, but just wanted to be sure. 🤔

rastislavvasko commented:
> You mean having all boolean fields as `isLoremIpsum` or all fields as `loremIpsum` ? If I understood correctly, it's the former, but just wanted to be sure. 🤔

The former (all boolean properties prefixed with `is`), yes!

> I'm asking because e.g. in Item, the property is already named `isChecked` instead of `checked` and `isDeleted` instead of `deleted`.

Huh, I thought it was just `Item.checked`, my bad. 🙃

rastislavvasko commented:
Everything looks good now!

pedrodanielcsantos commented:
Thanks a lot for the review once again! 🙏🥂 As agreed privately , I'll squash and merge now! 🚀
